hs_core/hydroshare/features.py

```diff
@@ -0,0 +1,434 @@
+        return records
+
+    @staticmethod
+    def group_editors():
+        """ return (user, group) tuples representing group editing privilege """
+        records = []
+        for u in User.objects.all():
+            editable = Group.objects.filter(g2ugp__user=u,
+                                            g2ugp__privilege__lte=PrivilegeCodes.CHANGE)
+            for g in editable:
+                records.append((u.username, g.name,))
+        return records
+
+    @staticmethod
+    def resource_viewers(fromdate, todate):
+        """ map of users who viewed each resource, according to date of access """
+        expr = re.compile('/resource/([^/]+)/')  # home page of resource
+        resource_visited_by_user = {}
+        for v in Variable.objects.filter(name='visit',
+                                         timestamp__gte=fromdate, timestamp__lte=todate):
+            user = v.session.visitor.user
+            if user is not None and \
+               user.username != 'test' and user.username != 'demo':
+                value = v.get_value()
+                m = expr.search(value)  # home page of resource
+                if m and m.group(1):
+                    resource_id = m.group(1)
+                    user_id = user.username
+                    # print("user={} resource={} when={}".format(user_id, resource_id,
+                    #     str(v.timestamp)))
+                    if resource_id not in resource_visited_by_user:
+                        resource_visited_by_user[resource_id] = set([user_id])
+                    else:
+                        resource_visited_by_user[resource_id].add(user_id)
+        return resource_visited_by_user
+
+    @staticmethod
+    def visited_resources(fromdate, todate):
+        """ map of users who viewed each resource, according to date of access """
+        expr = re.compile('/resource/([^/]+)/')  # home page of resource
+        user_visiting_resource = {}
+        for v in Variable.objects.filter(name='visit',
+                                         timestamp__gte=fromdate, timestamp__lte=todate):
+            user = v.session.visitor.user
+            if user is not None and \
+               user.username != 'test' and user.username != 'demo':
+                value = v.get_value()
+                m = expr.search(value)  # home page of resource
+                if m and m.group(1):
+                    resource_id = m.group(1)
+                    user_id = user.username
+                    # print("user={} resource={} when={}".format(user_id, resource_id,
+                    #     str(v.timestamp)))
+                    if user_id not in user_visiting_resource:
+                        user_visiting_resource[user_id] = set([resource_id])
+                    else:
+                        user_visiting_resource[user_id].add(resource_id)
+        return user_visiting_resource
+
+    @staticmethod
+    def resource_downloads(fromdate, todate):
```

Comment:
need a comment describing what this does. It should define meanings and types of parameters. Are they strings or dates?

Reply:
Resolved